src/xb-database-manager.c

```diff
@@ -352,6 +352,79 @@ xb_database_manager_register_stopwords (XbDatabaseManager *self,
   return TRUE;
 }
 
+static XapianDatabase *
+create_database_from_manifest (const char  *manifest_path,
+                               GError     **error_out)
+{
+  JsonParser *parser;
+  JsonNode *node;
+  JsonObject *json_manifest;
+  JsonArray *json_dbs;
+  GList *dbs, *l;
+  XapianDatabase *db = NULL;
+  GError *error = NULL;
+  g_autofree char *manifest_dir_path = NULL;
+
+  parser = json_parser_new ();
+  if (!json_parser_load_from_file (parser, manifest_path, &error))
+    {
+      g_propagate_error (error_out, error);
+      goto out;
+    }
+
+  db = xapian_database_new (&error);
+  if (error != NULL)
+    {
+      g_propagate_error (error_out, error);
+      goto out;
+    }
+
+  node = json_parser_get_root (parser);
+  json_manifest = json_node_get_object (node);
+  json_dbs = json_object_get_array_member (json_manifest, "xapian_databases");
+  dbs = json_array_get_elements (json_dbs);
+
+  manifest_dir_path = g_path_get_dirname (manifest_path);
+
+  for (l = dbs; l != NULL; l = l->next)
+    {
+      JsonObject *json_db = json_node_get_object (l->data);
+
+      int db_offset = json_object_get_int_member (json_db, "offset");
+
+      const char *relpath = json_object_get_string_member (json_db, "path");
+      g_autofree char *db_path = g_build_filename (manifest_dir_path, relpath, NULL);
+
+      XapianDatabase *internal_db = g_initable_new (XAPIAN_TYPE_DATABASE,
+                                                    NULL, &error,
+                                                    "path", db_path,
+                                                    "offset", db_offset,
+                                                    NULL);
+
+      if (error != NULL)
+        {
+          g_propagate_error (error_out, error);
+          goto out;
+        }
+
+      xapian_database_add_database (db, internal_db);
+    }
+
+ out:
+  g_clear_object (&parser);
+  return db;
+}
+
+static const char *
+xb_database_path (XbDatabase xbdb)
```

Comment:
`path` is for old-style Xapian databases, e.g. ones in V1 or V2 domains, with the `postlist.db` file and everything. It could also be for single-file databases, but not embedded in shards, since we don't support offsets.

`manifest_path` is for multi-shard databases.

We need both for compatibility reasons unless we want to generate a manifest for V1 / V2 bundles as well.